x-pack/plugins/security_solution/public/cases/attachments/types.ts

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ResponseActionAgentType } from '../../../common/endpoint/service/response_actions/constants';
+
+export interface IExternalReferenceMetaDataProps {
+  externalReferenceMetadata: {
+    comment: string;
+    command: string;
+    targets: Array<{
+      endpointId: string;
+      hostname: string;
+      agentType: ResponseActionAgentType;
+    }>;
+  };
+}
```

Comment:
Why do we have this `externalReferenceMetadata` indirection? If it does not have any special purpose, would it be possible to define flat component props as usual? like:

```suggestion
export interface IExternalReferenceMetaDataProps {
  comment: string;
  command: string;
  targets: Array<{
    endpointId: string;
    hostname: string;
    agentType: ResponseActionAgentType;
  }>;
}
```

Reply:
Hey, thanks for the comment 👍  
externalReferenceMetadata is what comes along with other props when using both `getAttachmentViewObject` and `children`. I could destructure that if you prefer, but would it increase readability? What do you think?
```
 getAttachmentViewObject: (props: IExternalReferenceMetaDataProps) => {
      const iconType = props.externalReferenceMetadata?.command === 'isolate' ? 'lock' : 'lockOpen';
      return {
        type: 'regular',
        event: getLazyExternalEventContent(props),
        timelineAvatar: (
          <EuiAvatar name="endpoint" color="subdued" iconType={iconType} aria-label={iconType} />
        ),
        children: getLazyExternalChildrenContent,
      };
    },
    ```

Reply:
Okay I see. No, that's okay if the same props are passed through the compoent tree. But, to increase readability, instead of redefining each component props type with a subset of the props we are really passing, like:

```
const AttachmentContentEvent = (props: {
  externalReferenceMetadata: {
    command: IExternalReferenceMetaDataProps['externalReferenceMetadata']['command'];
    targets: IExternalReferenceMetaDataProps['externalReferenceMetadata']['targets'];
  };
}) => {
  const {
    externalReferenceMetadata: { command, targets },
  } = props;
[...]
```

It would make sense to set the entire type we pass, and destruct only the relevant props from the function param:
```
const AttachmentContentEvent = ({
  externalReferenceMetadata: { command, targets },
}: IExternalReferenceMetaDataProps) => {
[...]
```

Reply:
Thanks @semd I think I understand what you're suggesting, but let me tell you why I do not feel good about this approach :P 

Approach 1: 
```
const AttachmentContentEvent = ({
  externalReferenceMetadata: { command, targets },
}: IExternalReferenceMetaDataProps) => {
[...]
```

By using the entire IExternalReferenceMetaDataProps, we are being more permissive and expecting all the properties of the interface to be present, which can lead to unintended dependencies and potential issues if the interface changes in the future. In short words, we expect props that we do not actually use.

Approach 2: 
```
const AttachmentContentEvent = (props: {
  externalReferenceMetadata: {
    command: IExternalReferenceMetaDataProps['externalReferenceMetadata']['command'];
    targets: IExternalReferenceMetaDataProps['externalReferenceMetadata']['targets'];
  };
}) => {
```

The second approach with more specific typings for the properties we actually use, is a form of `covariance`, where we are specifying exactly what you need and nothing more. This can make our code more robust, easier to understand, and less prone to errors when changes occur. In short words, we can use it in many places with many parent interfaces as long as the expected minimum is met. 

However, I think it makes sense to change my strangely looking:
`command: IExternalReferenceMetaDataProps['externalReferenceMetadata']['command'];`
into:
`command: ExternalReferenceCommand` 

Let me know what you think about it :) Thanks again! ❤️

Reply:
Hi @tomsonpl I like your explanation, I appreciate the reasoning behind the approach. But since we are sharing opinions I'll give mine :)

I think these components are very particular to fit this specific use case, they are not supposed to be reused somewhere else, so I would prime simplicity and consistency over reusability: 

Since we always have the entire `IExternalReferenceMetaData` object available and it seems we pass the whole object to all components by convention, I would reflect that in the components types, this will make all the properties available to them by type, they are all always passed so they are available on runtime, reflecting this fact in the type would make it more consistent with what we receive. And it's not a problem to have unused parts of the props, it's only typing. 

This approach would be as robust as the other one as long as we only use the components here, which I think is fair to assume. But, I agree with you that if we have to use them somewhere else, and we don't always have the entire `IExternalReferenceMetaData` object available, we would have to create it with a bunch of properties that are not used by the component, so that would not be a good design. 

Nevertheless, if I were to make these components reusable, I would not assume the prop shape would contain the `externalReferenceMetadata` indirection, it would define only what it needs, without making any assumptions of the context in which the component will be placed, so I would define the props with just:
```
const AttachmentContentEvent = (props: {
  command: ExternalReferenceComment;
  targets: ExternalReferenceTargets;
}) => {
```

However, as I said, if these components have not been thought to be reusable from the beginning, I would not worry about that.
That's my opinion, but no need to change anything. I already approved the PR, so this topic is not something that worries me at all, it's okay the way you have it right now 👍

Reply:
Thanks @semd, I agree, your explanation makes sense to me, since these components indeed are unlikely to be used somewhere else :) 
I'll adjust the code to your version with the next commit (if it happens :P) 

Thanks for the discussion, really appreciate some thoughts exchange :)